arangod/Aql/OptimizerRulesReplaceFunctions.cpp

```diff
@@ -0,0 +1,464 @@
+    expressionAst = ast->createNodeBinaryOperator(
+      AstNodeType::NODE_TYPE_OPERATOR_BINARY_LE, funDist ,params.radius
+    );
+  }
+
+  //// create calculation node used in SORT or FILTER
+  //Calculation Node will acquire ownership
+  Expression* calcExpr = new Expression(plan, ast, expressionAst);
+
+  // put condition into calculation node
+  Variable* calcOutVariable = ast->variables()->createTemporaryVariable();
+  ExecutionNode* eCalc = plan->registerNode(
+      new CalculationNode(plan, plan->nextId(), calcExpr, nullptr, calcOutVariable)
+  );
+  eCalc->addDependency(eEnumerate);
+
+  //// create SORT of FILTER
+  ExecutionNode* eSortOrFilter = nullptr;
+  if(isNear){
+    // use calculation node in sort node
+    SortElementVector sortElements { SortElement{ calcOutVariable, /*asc*/ true} }; //CHECKME
+    eSortOrFilter = plan->registerNode(
+        new SortNode(plan, plan->nextId(), sortElements, false)
+    );
+  } else {
+    eSortOrFilter = plan->registerNode(
+        new FilterNode(plan, plan->nextId(), calcOutVariable)
+    );
+  }
+  eSortOrFilter->addDependency(eCalc);
+
+  //// create MERGE(d, { param.distname : DISTANCE(d.lat, d.long, param.lat, param.lon)})
+  if(params.distanceName) { //return without merging the distance into the result
+    AstNode* elem = nullptr;
+    AstNode* funDistMerge = nullptr;
+    if(isNear){
+      funDistMerge = ast->createNodeReference(calcOutVariable);
+    } else {
+      //FIXME: recycles Ast directly - must probably be fixed;
+      funDistMerge = funDist;
+    }
+    if(params.distanceName->isConstant()){
+      elem = ast->createNodeObjectElement(params.distanceName->getStringValue()
+                                         ,params.distanceName->getStringLength()
+                                         ,funDistMerge);
+    } else {
+      elem = ast->createNodeCalculatedObjectElement(params.distanceName, funDistMerge );
+    }
+    auto* obj = ast->createNodeObject();
+    obj->addMember(elem);
+
+    auto* argsArrayMerge = ast->createNodeArray();
+    argsArrayMerge->addMember(docRef);
+    argsArrayMerge->addMember(obj);
+
+    auto* funMerge = ast->createNodeFunctionCall(TRI_CHAR_LENGTH_PAIR("MERGE"), argsArrayMerge);
+
+    Variable* calcMergeOutVariable = ast->variables()->createTemporaryVariable();
+    Expression* calcMergeExpr = new Expression(plan, ast, funMerge);
+    ExecutionNode* eCalcMerge = plan->registerNode(
+      new CalculationNode(plan, plan->nextId(), calcMergeExpr, nullptr, calcMergeOutVariable)
+    );
+    plan->insertAfter(eSortOrFilter, eCalcMerge);
+
+    //// wrap plan part into subquery
+    return createSubqueryWithLimit(plan, calcNode, eEnumerate, eCalcMerge, calcMergeOutVariable, params.limit);
+  }
+
+  //// wrap plan part into subquery
+  return createSubqueryWithLimit(plan, calcNode,
+                                eEnumerate /* first */, eSortOrFilter /* last */,
+                                enumerateOutVariable, params.limit);
+}
```

Comment:
This function is super-long. Is there a good way to break it up? I had to re-read several times because I forgot what was going on already by the time I got near the end of it...

Reply:
It has 40 lines. Imho there are other functions that need splitting.

Reply:
Sorry, I was referring to `replaceNearOrWithin`, which starts at line 190 and ends there at 354.

Reply:
I have shortened that function a bit by pulling out the index related code. Maybe the `MERGE` code could go in another function but that uses quite a few local (around 8) variables. That is why I did not split it further up.

Reply:
Okay, that helps already. Fine by me.